static/src/javascripts-legacy/projects/common/utils/check-mediator.js

```diff
@@ -0,0 +1,146 @@
+                list:[{
+                    id: 'isUserInAClashingAbTest'
+                }, {
+                    id: 'canEmailBeInserted',
+                    dependentChecks: {
+                        passCondition: EVERYCHECKPASSED,
+                        list: [{
+                            id: 'emailCanRun'
+                        }, {
+                            id: 'listCanRun'
+                        }, {
+                            id: 'emailInArticleOutbrainEnabled'
+                        }]
+                    }
+                }]
+            }
+        }];
+
+    function DeferredCheck(dependentCheckPromises, dependentChecksPassCondition) {
+        this.complete = new Promise(function(resolve, reject) {
+            this.resolve = resolve;
+            this.reject = reject;
+        }.bind(this));
+
+        if (dependentCheckPromises) {
+            Promise.all(dependentCheckPromises).then(function(results) {
+
+                var hasPassed = function(result) {
+                    return result;
+                };
+
+                if (dependentChecksPassCondition.call(results, hasPassed)) {
+                    this.resolve(true);
+                } else {
+                    this.resolve(false);
+                }
+            }.bind(this));
+        }
+    }
+
+    function registerDefferedCheck(check) {
+        var dependentCheckPromises = [];
+
+        if (check.dependentChecks) {
+            check.dependentChecks.list.forEach(registerDependentCheck.bind(null, dependentCheckPromises));
+        
+            return new DeferredCheck(dependentCheckPromises, check.dependentChecks.passCondition);
+        }
+
+        return new DeferredCheck();
+    }
+
+    
+    function registerDependentCheck(dependentCheckPromises, dependentCheck) {
+        dependentCheckPromises.push(registerCheck(dependentCheck).complete);
+    }    
+
+    function registerCheck(check) {
+        var registeredCheck = registerDefferedCheck(check);
+
+        registeredChecks[check.id] = registeredCheck;
+
+        return registeredCheck;
+    }
+
+    /**
+     * public
+    **/
+    function init() {
+        checkList.forEach(registerCheck);
+    }
+    
+    function resolveCheck(id) {
+        var argsArray = Array.prototype.slice.call(arguments, 1);
+        
+        if (registeredChecks[id]) {
+            return registeredChecks[id].resolve.apply(null, argsArray);
+        }
+    }
+
+    function rejectCheck(id) {
+        var argsArray = Array.prototype.slice.call(arguments, 1);
+
+        if (registeredChecks[id]) {
+            return registeredChecks[id].reject.apply(null, argsArray);
+        }
+    }
+
+    function waitForCheck(id) {
+        if (registeredChecks[id]) {
+            return registeredChecks[id].complete;
+        }
+
+        return Promise.reject('no deferred check with id ' + id);
+    }
+
+    /**
+     * exposed for unit testing
+    **/
+    function _testRegisterCheck(check) {
```

Comment:
Yes I did consider this, but at the time it felt messy defining the checkList var else where (in main.js) and passing it into checkMediator's init, because checkList has the potential to grow. In terms of implementation though I agree it'd be neater to pass this to checkMediator as then we can do away with the test functions.